internal/characters/layla/skill.go

```diff
@@ -0,0 +1,86 @@
+package layla
+
+import (
+	"github.com/genshinsim/gcsim/internal/frames"
+	"github.com/genshinsim/gcsim/pkg/avatar"
+	"github.com/genshinsim/gcsim/pkg/core/action"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/event"
+	"github.com/genshinsim/gcsim/pkg/core/player/shield"
+)
+
+var skillFrames []int
+
+const (
+	skillEnergy = "skill-energy-icd"
+
+	skillHitmark = 32
+)
+
+func init() {
+	skillFrames = frames.InitAbilSlice(43) // E -> Q/J
+	skillFrames[action.ActionAttack] = 41  // E -> N1
+	skillFrames[action.ActionSkill] = 42   // E -> E
+	skillFrames[action.ActionDash] = 43    // E -> D
+	skillFrames[action.ActionWalk] = 42    // E -> W
+	skillFrames[action.ActionSwap] = 41    // E -> Swap
+}
+
+func (c *char) Skill(p map[string]int) action.ActionInfo {
+	ai := combat.AttackInfo{
+		ActorIndex: c.Index,
+		Abil:       "Nights of Formal Focus",
+		AttackTag:  combat.AttackTagElementalArt,
+		ICDTag:     combat.ICDTagNone,
+		ICDGroup:   combat.ICDGroupDefault,
+		StrikeType: combat.StrikeTypeDefault,
+		Element:    attributes.Cryo,
+		Durability: 25,
+		Mult:       skill[c.TalentLvlSkill()],
+	}
+	c.Core.QueueAttack(ai, combat.NewCircleHit(c.Core.Combat.Player(), 2.5), 0, skillHitmark)
+
+	// add shield
+	c.QueueCharTask(func() {
+		exist := c.Core.Player.Shields.Get(shield.ShieldLaylaSkill)
+		if exist == nil {
+			shield := shieldBase[c.TalentLvlSkill()] + shieldPer[c.TalentLvlSkill()]*c.MaxHP()
+			c.Core.Player.Shields.Add(c.newShield(shield, 12*60))
+		} else {
+			shd, _ := exist.(*shield.Tmpl)
+			shd.Expires = c.Core.F + 12*60
+		}
+	}, 6)
+
+	// apply cryo & run a task
+	c.QueueCharTask(func() {
+		player, ok := c.Core.Combat.Player().(*avatar.Player)
+		if !ok {
+			panic("target 0 should be Player but is not!!")
+		}
+		player.ApplySelfInfusion(attributes.Cryo, 25, 0.1*60)
+
+		c.starTickSrc = c.Core.F
+		c.TickNightStar(c.starTickSrc, false)
+	}, 32)
+
+	c.SetCDWithDelay(action.ActionSkill, 12*60, 19)
+
+	return action.ActionInfo{
+		Frames:          frames.NewAbilFunc(skillFrames),
+		AnimationLength: skillFrames[action.InvalidAction],
+		CanQueueAfter:   skillFrames[action.ActionDash], // earliest cancel is before skillHitmark
+		State:           action.SkillState,
+	}
+}
+
+func (c *char) StarsSkill() {
```

Comment:
Capital letter in the beginning changes visibility, so please rename it to starsSkill.